### PR TITLE
fix: Adding support for default path of configfile #363

### PR DESCRIPTION
**Pull request by KrGolovin**
also adding test for completion and resolve

### CODE REVIEW

**Comment by KrGolovin** [comment]
configfile

File: src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkFileReference.kt [hunk 1]
```diff
@@ -158,17 +179,26 @@ class SmkIncludeReference(
...
 ) : SmkFileReference(element, textRange, stringLiteralExpression, path) {
-    override fun getVariants() =  collectFileSystemItemLike() {
+    override fun getVariants() =  collectFileSystemItemLike(
+        onlyFromParentFolder = false,
+        isFromCurrentDir = false
+    ) {
         isYamlFile(it)
     }
+
+    override fun resolve(): PsiElement? = findPathToResolve(searchFromRoot = true)
```
Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkFileReference.kt:197]: At the moment we have 3 flags: `onlyFromParentFolder`, `isFromCurrentDir`, `searchFromRoot`. At the moment the difference is not clear and no documentation available. I assume that `onlyFromParentFolder`, `isFromCurrentDir` is same thing. Could you refactor this code:
* convert this fields to `SmkFileReference` class abstract variables
* rename them to more clear names + add documentation. I suppose good names could be `searchRelativelyToContentRoot`, `searchRelativelyToCurrentFolder`, where current folder is parent folder of current snakemake file, and content root is the snakemake project working directory

File: src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkFileReference.kt [hunk 2]
```diff
@@ -100,28 +102,45 @@ open class SmkFileReference(
...
-    private fun findVirtualFile(): VirtualFile? {
+    private fun findVirtualFile(searchFromRoot: Boolean = false): VirtualFile? {
         val stringLiteral = element.argumentList?.arguments?.firstOrNull {
             it is PyStringLiteralExpression && it.stringValue == path
         } ?: return null
 
-        val parentFolder = element.containingFile.originalFile.virtualFile?.parent ?: return null
-
-        val filePath = (stringLiteral as? PyStringLiteralExpression)?.stringValue
+        val filePath = (stringLiteral as? PyStringLiteralExpression)?.stringValue!!
```
Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkFileReference.kt:110]: Not clear how it works `(stringLiteral as? PyStringLiteralExpression)?` is marked as nullable, returns `null` or expression, but than is declared as not-null `!!`

Comment (iromeo) [reply src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkFileReference.kt:110]: Moreover prev line always returns `PyStringLiteralExpression` with `stringValue == path`, so there is no need to calculate `stringValue` again. Type checker fails to infer that stringLiteral is `PyStringLiteralExpression` from the above code, but we could refactor it using:

```kotline
        val stringLiteral = element.argumentList?.arguments?.firstOrNull { 
                it is PyStringLiteralExpression &&  it.stringValue == path 
            } as PyStringLiteralExpression? ?: return null

        // or
        val stringLiteral = element.argumentList?.arguments
            ?.filterIsInstance(PyStringLiteralExpression::class.java)
            ?.firstOrNull { it.stringValue == path }
            ?: return null
        
// in both cases `stingLiteral` type will be inferred as `PyStringLiteralExpression`
            
```

Comment (iromeo) [reply src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkFileReference.kt:110]: Also, is not clear why we need `stringLiteral` check in the first line of the method, seems could be refactored just to:
```kotlin
    private fun findVirtualFile(searchFromRoot: Boolean = false): VirtualFile? {
        // Try to find using relative path
        val relativeFile = if (searchFromRoot) {
            //search in all content roots
            ProjectRootManager.getInstance(element.project).contentRoots.firstOrNull { root ->
                root.findFileByRelativePath(path) != null
            }
        } else {
            element.containingFile.originalFile.virtualFile?.parent
                ?.findFileByRelativePath(path)
        }
        if (relativeFile != null) {
            return relativeFile
        }
        // Trying to find the file anywhere
        val vfm = VirtualFileManager.getInstance()
        val localFS = vfm.getFileSystem(LocalFileSystem.PROTOCOL)
        return localFS.findFileByPath(path) ?: vfm.findFileByUrl(path)
    }
```

**Review by iromeo — COMMENTED** [review]
Feature works, but code needs to be refactored + some tests are missing

File: src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkFileReference.kt (hunk 1 above)
Comment (KrGolovin) [reply src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkFileReference.kt:197]: We have for example "conda:" field,  where we search file from root, but path is making from current folder:
"../envs/snpsift.yaml"
Therefore, we need 2 flags to search for a file and build a path

File: src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkFileReference.kt (hunk 2 above)
Comment (KrGolovin) [reply src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkFileReference.kt:110]: Ok, this is really better, but we have to use firstNotNullOfOrNull

File: src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkFileReference.kt (hunk 1 above)
Comment (KrGolovin) [reply src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkFileReference.kt:197]: I don't understand a bit what an abstract variable is, so for now I have made it the usual fields of the class

Comment (iromeo) [reply src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkFileReference.kt:197]: I meant of abstract property:
```kotlin
abstract class SmkFileReference(..) {
   abstract val  searchRelativelyToCurrentFolder: Boolean
}
class Foo(): SmkFileReference(..) {
  override val searchRelativelyToCurrentFolder = True
} 
```
but your choice (private property + initialization via constructor) is also a good choice. Let's leave it as is.

Comment (iromeo) [reply src/main/kotlin/com/jetbrains/snakecharm/lang/psi/SmkFileReference.kt:197]: > We have for example "conda:" field
Discussion moved to the new PR.

**Pull request closed without merging**